[Public/Global - Microsoft Azure Attestation Service] Adding September 2022 preview Microsoft Attestation API version

yinqchen opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
Adding one new APIs to support 
- Retrieves all available Tcb baselines for an attestation type

Updates:
1. Added `"format": "date-time"` to `tcbReleaseDate`
2. Removed query parameter `provideCurrentTcbInPolicy`. Added `isSelectedTcb` property with boolean type to indicate the baseline is current selected/set baseline in attestation policy

We had some discussion within our team:
3. Instead of adding a property `isAzureDefault` boolean, added a property `tcbIdentifier` with string type because
- In order to use a specified tcb, customer needs to set the tcb in attestation policy (see an example below). The property name returns from GetTcbBaseline API should match the tcb configuration in policy. We were using 
`issueproperty(type = "x-ms-sgx-tcbevaluationdatanumber", value = "azuredefault")` in attestation policy. We agree that `azuredefault` is not a number, using `tcbIdentifier` is more appropriated. 
- `azuredefault` tcb will be updated periodically by Internal Azure team. In the near future, we may introduce another baseline `sqldefault`; this baseline is specifically for _Always Encrypted SQL_ customers. It will be updated periodically by Internal Azure team on a different schedule. With this change, we don't need to update swagger every time we introduce an internally managed auto updated tcb baseline

Attestation policy example:
![image](https://user-images.githubusercontent.com/40009687/195798278-9edf1209-c125-4221-8a76-ed2baf0c25e8.png)

Customer can also use:
`issueproperty(type = "x-ms-sgx-tcbIdentifier", value = "10")`
or 
`issueproperty(type = "x-ms-sgx-tcbIdentifier", value = "11")`
if they would like to pin to a specific tcb version.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/20922

yinqchen committed:
rest api update for tcb baseine

yinqchen committed:
address pr comments

yinqchen committed:
update examples

LarryOsterman commented on specification/attestation/data-plane/Microsoft.Attestation/preview/2022-09-01-preview/attestation.json:
```diff
@@ -0,0 +1,1206 @@
+          "type": "string"
+        },
+        "tcbReleaseDate": {
+          "description": "Date and time when the Tcb is released. SGX TEE specific property",
```
If this is a string, how do you reflect the customers time format preferences? Would it make sense to make this a [`"format":"dateTime"`](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md#data-types) field?

heaths replied:
Clarification: `"format": "date-time"`.

yinqchen replied:
Thanks! Updated.

heaths commented on specification/attestation/data-plane/Microsoft.Attestation/preview/2022-09-01-preview/attestation.json:
```diff
@@ -0,0 +1,1206 @@
+        }
+      }
+    },
+    "TcbBaselineResult": {
```
Seems this list could grow a lot over time. This should probably be pageable, which means adding an [`x-ms-pageable`](https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-pageable) to the operation returning it, and a `nextLink` to the properties of this object. Even if you don't page now, if it seems even _remotely_ possible in the future, you should now or this will be a breaking change later.

LarryOsterman replied:
There will be one entry added every 6 months. Over a decade, that means adding 20 entries to the list. Given that, does it still make sense to make it pagable?

heaths replied:
That's really for the service team to say. My only cautionary example from back in the VS setup days when we worked together a bit: never say never. Too many teams have said, "that will never happen" and years (or even months) later it did and they often didn't have a lot of options. If there's a remote change this list could grow large, I'd support paging now even if you don't use it initially. If you're absolutely sure it never will be a long list, it's your call.

heaths replied:
From the meeting, it sounded like this wasn't likely (as you said), so resolving.

yinqchen committed:
address pr comments

yinqchen committed:
add format date-time for tcbrelease date

yinqchen commented:
1. Added `"format": "date-time"` to `tcbReleaseDate`
2. Removed query parameter `provideCurrentTcbInPolicy`. Added `isSelectedTcb` property with boolean type to indicate the baseline is current selected/set baseline in attestation policy

We had some discussion within our team:
3. Instead of adding a property `isAzureDefault` boolean, added a property `tcbIdentifier` with string type because
- In order to use a specified tcb, customer needs to set the tcb in attestation policy (see an example below). The property name returns from GetTcbBaseline API should match the tcb configuration in policy. We were using 
`issueproperty(type = "x-ms-sgx-tcbevaluationdatanumber", value = "azuredefault")` in attestation policy. We agree that `azuredefault` is not a number, using `tcbIdentifier` is more appropriated. 
- `azuredefault` tcb will be updated periodically by Internal Azure team. In the near future, we may introduce another baseline `sqldefault`; this baseline is specifically for _Always Encrypted SQL_ customers. It will be updated periodically by Internal Azure team on a different schedule. With this change, we don't need to update swagger every time we introduce an internally managed auto updated tcb baseline

Attestation policy example:
![image](https://user-images.githubusercontent.com/40009687/195798278-9edf1209-c125-4221-8a76-ed2baf0c25e8.png)

Customer can also use:
`issueproperty(type = "x-ms-sgx-tcbIdentifier", value = "10")`
or 
`issueproperty(type = "x-ms-sgx-tcbIdentifier", value = "11")`
if they would like to pin to a specific tcb version.

yinqchen committed:
remove space

yinqchen committed:
update format

yinqchen committed:
update description

yinqchen commented:
@heaths @BlackRider97 @rysweet Can someone please take a look at this PR and approve it?

heaths commented on specification/attestation/data-plane/Microsoft.Attestation/preview/2022-09-01-preview/attestation.json:
```diff
@@ -467,22 +467,14 @@
           "TcbBaseline"
         ],
-        "summary": "Retrieves the current set Tcb baselines or all available Tcb baselines for an attestation type.",
+        "summary": "Retrieves a list of all azure supported baseline details for the a TEE type along with an indication of which one is in enforced for the attestation provider.",
```
We should properly case our own product names.
```suggestion
        "summary": "Retrieves a list of all Azure supported baseline details for the a TEE type along with an indication of which one is in enforced for the attestation provider.",
```

heaths replied:
Several other instances below.

yinqchen replied:
Thanks! Updated

heaths commented on specification/attestation/data-plane/Microsoft.Attestation/preview/2022-09-01-preview/attestation.json:
```diff
@@ -748,6 +743,10 @@
           "description": "Minimum Windows PSW version required to support the corresponding Tcb baseline. SGX TEE specific property",
           "type": "string"
+        },
+        "isSelectedTcb": {
```
Wasn't there to be a change to the endpoint to get _only_ the selected TCB?

yinqchen replied:
Per API review discussion, we agreed upon not having an endpoint to get only the selected TCB. Customers will go through the list and find the selected TCB.
We will create an endpoint to get selected TCB if there is customer request.

yinqchen committed:
update case for azure

yinqchen committed:
Merge branch 'Azure:main' into attest-tcb
